responses.go

```diff
@@ -32,3 +36,53 @@ func (r *SortResponse) Handle(resp imap.Resp) error {
 func (r *SortResponse) WriteTo(w *imap.Writer) error {
 	return errors.New("sortthread: not yet implemented")
 }
+
+func (r *ThreadResponse) Handle(resp imap.Resp) error {
+	name, fields, ok := imap.ParseNamedResp(resp)
+	if !ok || name != "THREAD" {
+		return responses.ErrUnhandled
+	}
+	if threads, err := parseThreadResp(fields); err != nil {
+		return err
+	} else {
+		r.Threads = threads
+	}
+
+	return nil
+}
+
+func parseThreadResp(fields []interface{}) ([]*Thread, error) {
+	var parent *Thread
+	var siblings []*Thread
+	for _, f := range fields {
+		switch f := f.(type) {
+		case string:
+			id, err := imap.ParseNumber(f)
+			if err != nil {
+				return nil, err
+			}
+			t := Thread{Id: id}
+			if parent == nil {
+				siblings = append(siblings, &t)
+			} else {
+				parent.Children = append(t.Children, &t)
+			}
+			parent = &t
+		case []interface{}:
+			t, err := parseThreadResp(f)
```

Comment:
Lists are parsed differently when they are children and when they are toplevel. For instance:

```
1 2 3
```

This is three threads.

```
(1 2 3)
```

This is a single thread. 1 has a single child: 2. 2 has a single child: 3.

See the RFC example.

Reply:
The first example you gave would look like this (1)(2)(3). 

I also have tests for both of these cases. Though I agree the tests need to be reworked as it is not easy to grok what is going on.